File: packages/typescript/src/features/advanced-capabilities/actions/addContact.ts
import { findEntityByName } from "../../../entities.ts";
import { requireActionSpec } from "../../../generated/spec-helpers.ts";
import {
	findKeywordTermMatch,
	getValidationKeywordTerms,
} from "../../../i18n/validation-keywords.ts";
import { logger } from "../../../logger.ts";
import { addContactTemplate } from "../../../prompts.ts";
import type { RelationshipsService } from "../../../services/relationships.ts";
import type {
	Action,
	ActionExample,
	ActionResult,
	Entity,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
} from "../../../types/index.ts";
import { asUUID, ModelType } from "../../../types/index.ts";
import {
	composePromptFromState,
	parseKeyValueXml,
	stringToUuid,
} from "../../../utils.ts";

// Get text content from centralized specs
const spec = requireActionSpec("ADD_CONTACT");
const ADD_KEYWORDS = getValidationKeywordTerms("action.addContact.request", {
	includeAllLocales: true,
});

interface AddContactXmlResult {
	contactName?: string;
	entityId?: string;
	categories?: string;
	notes?: string;
	timezone?: string;
	language?: string;
	reason?: string;
}

export const addContactAction: Action = {
	name: spec.name,
	description: spec.description,
	similes: spec.similes ? [...spec.similes] : [],
	examples: (spec.examples ?? []) as ActionExample[][],

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
	): Promise<boolean> => {
		const relationshipsService = runtime.getService(
			"relationships",
		) as RelationshipsService;
		if (!relationshipsService) {
			logger.warn("[AddContact] RelationshipsService not available");
			return false;
		}

		const messageText = message.content.text ?? "";
		if (!messageText) return false;
		return findKeywordTermMatch(messageText, ADD_KEYWORDS) !== undefined;
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
	): Promise<ActionResult> => {
		const relationshipsService = runtime.getService(
			"relationships",
		) as RelationshipsService;

		if (!relationshipsService) {
			throw new Error("RelationshipsService not available");
		}

		if (!state) {
			state = {
				values: {},
				data: {},
				text: "",
			};
		}

		state.values = {
			...state.values,
			message: message.content.text,
			senderId: message.entityId,
			senderName: state.values?.senderName || "User",
		};

		const prompt = composePromptFromState({
			state,
			template: addContactTemplate,
		});

		const response = await runtime.useModel(ModelType.TEXT_SMALL, {
			prompt,
			stopSequences: [],
		});

		const parsedResponse = parseKeyValueXml<AddContactXmlResult>(response);
		if (!parsedResponse) {
			logger.warn(
				"[AddContact] Failed to parse contact information from response",
			);
			throw new Error("Could not extract contact information");
		}

		const contactName = parsedResponse.contactName?.trim();
		if (!contactName) {
			logger.warn("[AddContact] Missing contact name in response");
			throw new Error("Could not extract contact name");
		}

		let entityId = parsedResponse.entityId
			? asUUID(parsedResponse.entityId)
			: null;

		if (!entityId) {
			const entity = await findEntityByName(runtime, message, state);

			if (entity?.id) {
				entityId = entity.id;
			} else {
				// Create a new entity ID based on the name
				entityId = stringToUuid(`contact-${contactName}-${runtime.agentId}`);
				const entityToCreate: Entity = {
					id: entityId,
					names: [contactName],
					agentId: runtime.agentId,
				};
				await runtime.createEntity(entityToCreate);
			}
		}

		if (!entityId) {
			throw new Error("Could not determine entity ID for contact");
		}

		const existingEntity = await runtime.getEntityById(entityId);
		if (!existingEntity) {
			await runtime.createEntity({
				id: entityId,
				names: [contactName],
				agentId: runtime.agentId,
			});
		}

		// Parse categories
		const categories = parsedResponse.categories
			? parsedResponse.categories
					.split(",")
					.map((c: string) => c.trim())
					.filter(Boolean)
			: ["acquaintance"];

		// Build preferences
		const preferences: Record<string, string> = {};
		if (parsedResponse.timezone) preferences.timezone = parsedResponse.timezone;
		if (parsedResponse.language) preferences.language = parsedResponse.language;
		if (parsedResponse.notes) preferences.notes = parsedResponse.notes;

		const _contact = await relationshipsService.addContact(
			entityId,
			categories,
			preferences,
			{
				displayName: contactName,
			},
		);

		logger.info(`[AddContact] Added contact ${contactName} (${entityId})`);

		const responseText = `I've added ${contactName} to your contacts as ${categories.join(", ")}. ${
			parsedResponse.reason || "They have been saved to your relationships."
		}`;

		if (callback) {
			await callback({
				text: responseText,
				action: "ADD_CONTACT",
				metadata: {
					contactId: entityId,
					contactName,
					categories,
					success: true,
				},
			});
		}

		return {
			success: true,
			values: {
				contactId: entityId,
				contactName,
				categoriesStr: categories.join(","),
			},
			data: {
				contactId: entityId,
				contactName,
				categories: categories.join(","),
			},
			text: responseText,
		};
	},
};

File: packages/python/elizaos/features/advanced_capabilities/actions/add_contact.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_action_spec
from elizaos.prompts import ADD_CONTACT_TEMPLATE
from elizaos.types import (
    Action,
    ActionExample,
    ActionResult,
    Content,
    ModelType,
)
from elizaos.utils.spec_examples import convert_spec_examples
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import (
        HandlerCallback,
        HandlerOptions,
        IAgentRuntime,
        Memory,
        State,
    )

# Get text content from centralized specs
_spec = require_action_spec("ADD_CONTACT")


def _convert_spec_examples() -> list[list[ActionExample]]:
    """Convert spec examples to ActionExample format."""
    return convert_spec_examples(_spec)


@dataclass
class AddContactAction:
    name: str = _spec["name"]
    similes: list[str] = field(default_factory=lambda: list(_spec.get("similes", [])))
    description: str = _spec["description"]

    async def validate(
        self, runtime: IAgentRuntime, _message: Memory, _state: State | None = None
    ) -> bool:
        relationships_service = runtime.get_service("relationships")
        return relationships_service is not None

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        from elizaos.features.advanced_capabilities.services.relationships import (
            ContactPreferences,
            RelationshipsService,
        )

        relationships_service = runtime.get_service("relationships")
        if not relationships_service or not isinstance(relationships_service, RelationshipsService):
            return ActionResult(
                text="Relationships service not available",
                success=False,
                values={"error": True},
                data={"error": "RelationshipsService not available"},
            )

        state = await runtime.compose_state(message, ["RECENT_MESSAGES", "ENTITIES"])

        prompt = runtime.compose_prompt_from_state(
            state=state,
            template=ADD_CONTACT_TEMPLATE,
        )

        response = await runtime.use_model(ModelType.TEXT_SMALL, {"prompt": prompt})
        parsed = parse_key_value_xml(response)

        if not parsed or not parsed.get("contactName"):
            return ActionResult(
                text="Could not extract contact information",
                success=False,
                values={"error": True},
                data={"error": "Failed to parse contact info"},
            )

        contact_name = str(parsed.get("contactName", ""))
        categories_str = str(parsed.get("categories", "acquaintance"))
        categories = [c.strip() for c in categories_str.split(",") if c.strip()]
        notes = str(parsed.get("notes", ""))
        reason = str(parsed.get("reason", ""))

        from uuid import UUID as StdUUID

        entity_id = message.entity_id
        entity_id_uuid = StdUUID(str(entity_id)) if entity_id else None
        preferences = ContactPreferences(notes=notes) if notes else None

        if entity_id_uuid:
            await relationships_service.add_contact(
                entity_id=entity_id_uuid,
                categories=categories,
                preferences=preferences,
            )

        response_text = (
            f"I've added {contact_name} to your contacts as {', '.join(categories)}. {reason}"
        )

        if callback:
            await callback(Content(text=response_text, actions=["ADD_CONTACT"]))

        return ActionResult(
            text=response_text,
            success=True,
            values={
                "contactId": str(entity_id),
                "contactName": contact_name,
                "categoriesStr": ",".join(categories),
            },
            data={
                "contactId": str(entity_id),
                "contactName": contact_name,
                "categories": ",".join(categories),
            },
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return _convert_spec_examples()


add_contact_action = Action(
    name=AddContactAction.name,
    similes=AddContactAction().similes,
    description=AddContactAction.description,
    validate=AddContactAction().validate,
    handler=AddContactAction().handler,
    examples=AddContactAction().examples,
)
